### PR TITLE
squirrel: limit recursion for python lookup

### PR DESCRIPTION
**Pull request by mrnugget**
Customer [reported that `symbols` keeps crashing due to recursion in squirrel](https://sourcegraph.slack.com/archives/C03TGHH1S3V/p1666270245299469?thread_ts=1665750028.799609&cid=C03TGHH1S3V).

This is a quick fix for `symbols` to not crash. Instead it returns an "500 - internal server error" and logs the error.

It limits recursion in the Python "find def" call stack to 10000 call frames.

10000? Not sure about that, but seems like enough, since it only traverses a single document.

## Test plan

- Manual testing

## Testing video

https://user-images.githubusercontent.com/1185253/196975436-983cf6ba-b6f8-4e54-94f7-089a3537dd46.mp4

### CODE REVIEW

**Review by efritz — COMMENTED** [review]
This concerns me as you're creating a new context object wrapper on every call down the chain. These are probably small enough objects that they might not add up, but it could create a massive about of heap garbage in a short time.

Can we make the context reference a mutual object that increases a counter without allocation?

Or is this worrying over nothing?

**Comment by mrnugget** [comment]
> Can we make the context reference a mutual object that increases a counter without allocation?

I'll try

**Comment by mrnugget** [comment]
@chrismwendt Eric and I just changed the approach here from `context`-based to using squirrel's own `depth` value.

File: cmd/symbols/squirrel/service.go [hunk 1]
```diff
@@ -145,6 +145,7 @@ func (squirrel *SquirrelService) getDef(ctx context.Context, node Node) (*Node,
 	case "starlark":
 		return squirrel.getDefStarlark(ctx, node)
 	case "python":
+		fmt.Println("choo choo! all aboard the recursion train!")
```
Comment (efritz) [inline cmd/symbols/squirrel/service.go:148]:
```suggestion
```

**Review by efritz — APPROVED** [review]

File: cmd/symbols/squirrel/service.go [hunk 2]
```diff
@@ -172,11 +175,15 @@ func (squirrel *SquirrelService) onCall(node Node, arg fmt.Stringer, ret func()
 	squirrel.breadcrumbWithOpts(node, func() string { return msg }, 3)
 
 	squirrel.depth += 1
+	if squirrel.depth > maxGetDefStackDepth {
+		return nil, errors.New("max squirrel stack depth exceeded")
+	}
```
Comment (chrismwendt) [inline cmd/symbols/squirrel/service.go:180]: Could we panic here and recover in the HTTP handler? That way we don't need to litter `if err != nil` everywhere, right?

Comment (mrnugget) [reply cmd/symbols/squirrel/service.go]: I think this here is cleaner. I don't want to use `panic`/`recover` for control flow here.

**Review by chrismwendt — APPROVED** [review]

File: cmd/symbols/squirrel/service.go (hunk 2 above)
Comment (efritz) [reply cmd/symbols/squirrel/service.go]: @chrismwendt This also gives us the chance to say "we can try to find a def here, but ignore stack errors and move on to a default", which seems like a nice ability for the future. If we use panic/recover then there could be a substantial performance cost, and `recover` would need to be used outside of the top-level handler for control flow.

**Pull request merged**